### PR TITLE
better detection of malformed DD elements

### PR DESCRIPTION
**Pull request by gbirchmeier**
alternative impl for #592

@roederja2, what do you think?

### CODE REVIEW

**Comment by gbirchmeier** [comment]
Decided to do some needed refactoring and extra validity-checking while I was in here.

(Still hate the trinary `componentRequired` param of `ParseMsgEl()`, but I'll come back to that another time.)

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 1]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
+
+                switch(childNode.Name)
                 {
-                    if (FieldsByName.ContainsKey(fieldName) == false)
-                    {
-                        throw new Exception(
-                            $"Field '{fieldName}' is used in '{messageTypeName}', but is not defined in <fields> set.");
-                    }
+                    case "field":
+                        if (FieldsByName.ContainsKey(nameAttribute) == false)
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:602]:
```suggestion
                        if (!FieldsByName.ContainsKey(nameAttribute))
```

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 2]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
-                    grp.NumFld = fld.Tag;
-                    ParseMsgEl(childNode, grp);
-                    ddmap.Groups.Add(fld.Tag, grp);
+                        // if this is in a group whose delim is unset, then this must be the delim (i.e. first field)
+                        if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:619]:
```suggestion
                        if (ddmap is DDGrp ddGroup && ddGroup.Delim == 0)
                        {
                            ddGroup.Delim = fld.Tag;
                        }
```

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 3]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
-                        && (componentRequired == null || componentRequired.Value == true))
-                    {
-                        ddmap.ReqFields.Add(fld.Tag);
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
+                        DDField fld = FieldsByName[nameAttribute];
+                        if (required && (componentRequired == null || componentRequired.Value == true))
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:609]:
```suggestion
                        if (required && componentRequired.GetValueOrDefault(true))
```

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 4]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
-                    }
-                }
-                else if (childNode.Name == "component")
-                {
-                    XmlNode compNode = RootDoc.SelectSingleNode("//components/component[@name='" + fieldName + "']");
-                    XmlAttribute req = childNode.Attributes["required"];
-                    bool? compRequired = (req != null && req.Value == "Y");
-                    ParseMsgEl(compNode, ddmap, compRequired);
+                    case "group":
+                        if (FieldsByName.ContainsKey(nameAttribute) == false)
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:626]:
```suggestion
                        if (!FieldsByName.ContainsKey(nameAttribute))
```

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 5]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
-                        ((DDGrp)ddmap).Delim = fld.Tag;
-                    }
-                }
-                else if (childNode.Name == "component")
-                {
-                    XmlNode compNode = RootDoc.SelectSingleNode("//components/component[@name='" + fieldName + "']");
-                    XmlAttribute req = childNode.Attributes["required"];
-                    bool? compRequired = (req != null && req.Value == "Y");
-                    ParseMsgEl(compNode, ddmap, compRequired);
+                    case "group":
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:625]: There is a lot of copied code from the previous case. Could that be factored out into methods?

Comment (gbirchmeier) [reply QuickFIXn/DataDictionary/DataDictionary.cs:625]: I will definitely be coming back to this function later.

File: QuickFIXn/DataDictionary/DataDictionary.cs (hunk 2 above)
Comment (gbirchmeier) [reply QuickFIXn/DataDictionary/DataDictionary.cs:619]: never seen this syntax before...

File: QuickFIXn/DataDictionary/DataDictionary.cs (hunk 4 above)
Comment (gbirchmeier) [reply QuickFIXn/DataDictionary/DataDictionary.cs:626]: personal preference: I like the more visually-apparent `== false`

File: QuickFIXn/DataDictionary/DataDictionary.cs (hunk 1 above)
Comment (gbirchmeier) [reply QuickFIXn/DataDictionary/DataDictionary.cs:602]: personal preference: I like the more visually-apparent `== false`

File: QuickFIXn/DataDictionary/DataDictionary.cs (hunk 2 above)
Comment (roederja2) [reply QuickFIXn/DataDictionary/DataDictionary.cs]: Pattern matching - I think it was introduced with C# 7 - much safer than casting.

File: QuickFIXn/DataDictionary/DataDictionary.cs [hunk 6]
```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
...
+                        if (FieldsByName.ContainsKey(nameAttribute) == false)
+                        {
+                            throw new DictionaryParseException(
+                                $"Group counter field '{nameAttribute}' is not defined in <fields> section.");
+                        }
+
+                        DDField counterFld = FieldsByName[nameAttribute];
+                        DDGrp grp = new DDGrp();
+                        if (required && (componentRequired == null || componentRequired.Value == true))
```
Comment (roederja2) [inline QuickFIXn/DataDictionary/DataDictionary.cs:634]: This is now different than the `componentRequired.GetValueOrDefault(true)` used above...

Comment (gbirchmeier) [reply QuickFIXn/DataDictionary/DataDictionary.cs]: oops, missed that, fixed now.

**Comment by gbirchmeier** [comment]
Ok, I took one last pass at it and refactored the duplicate code out.

I think I'm done now.  If anyone still has comments please hit me with them, else I'll merge it.

**Pull request merged**